emscripten-simulated-jerry-api/jerry.js

```diff
@@ -148,6 +151,81 @@
       this._jerryToHostValueMap = {};
       this._nextObjectRef = 1;
       this._jerryInternalPropsWeakMap = new WeakMap();
+
+      if (!Module.Promise) {
+        console.warn('Please set Module.Promise to a promise implementation to use for JerryScript')
+      } else {
+        // Patch up the Promise.prototype then/catch implementations.
+        // The patched versions should create a function that when called, pushes the original callback arg to the jobQueue
+        var _then = Module.Promise.prototype.then;
+        var _catch = Module.Promise.prototype.catch;
+        var _identify = function(x) {return x};
+        var _thrower = function(x) {throw x};
+        var jobQueue = this.jobQueue;
+
+        // Modify the `then`.
+        Module.Promise.prototype.then = function (cb1, cb2) {
+          var res, rej;
+          var ret = new Module.Promise(function(f, r) {
+            res = f;
+            rej = r;
+          });
+          if (typeof cb1 !== 'function') {
+            cb1 = _identify;
+          }
+          if (typeof cb2 !== 'function') {
+            cb2 = _thrower;
+          }
+
+          var newCb1 = function(arg) {
+            jobQueue.push({arg: arg, cb: cb1, resolve:res, reject:rej});
+            return ret;
+          };
+
+          var newCb2 = function(arg) {
+            jobQueue.push({arg: arg, cb: cb2, resolve:res, reject:rej});
+            return ret;
+          };
+
+          return _then.call(this, newCb1, newCb2);
```

Comment:
I'd probably change this a little. Unlikely, but if the real `then` implementation uses `arguments.length` to check whether the 2nd, optional argument was passed, it will always receive `2` now.

So I'd instead write something like:
```
return _then.apply(this, [newCb1, newCb2].slice(0, arguments.length));
```